### PR TITLE
52/beginner friendly code examples

### PR DESCRIPTION
**Pull request by farmeroy**
I've tried to expand the view into the various files to give a more seemless code-along experience:
- guid the learner to create a `lib.rs` which includes the modules
- ensure that all imports are included and aliases are obvious
- clarify that the methods added to the `Tui` struct are within an implementation

To this end, I've also followed the example in the Rust Book of using `// --snip--` to show where code has been abbreviated.

It could be that this is a bit too didactic or hand holdy for your target audience, so these are just suggestions. 
I've been following along with the tutorial, and my personal style is to code along, typing out the code as we go. For other people doing the same, including these bits of code could be useful. They are really just sort of minor edits. I think this is a great tutorial and I'm super happy following along!

Happy to hear your thoughts

### CODE REVIEW

**Comment by kdheepak** [comment]
I think these kind of changes are great! Can you rebase on the latest `main` since I've made some modifications to these files as well. Also, I added a `.rustfmt.toml` file where it is 2 spaces for indent. We can later change it to 4 spaces if we want but for now it'll make the diffs less noisy.

**Comment by farmeroy** [comment]
So I rebased from main and then tried to rebase my remote branch from my local branch. Now, there is just a lot more noise from all the diffs between the main and my PR :( Sorry about that -- I haven't done much rebasing before, how can I clean this up for you?

**Comment by farmeroy** [comment]
Yeah, it looks I messed up the commit history. Somehow, it's added me as committer to some things other people have authored. I don't know what I did.  How can I fix it?

**Comment by joshka** [comment]
The following will get you back to where you were at:
Assuming that
- your local branch is also named "beginner-friendly-code-examples"
- your fork remote is named "origin"
- your ratatui-org/ratatui-book remote is named "upstream":

-1. If you want to understand how you got to where you were, take a look at `git reflog` before doing any of this, and perhaps after it as well - this will show you the transitions that git made when you did various commands.
0. If you have any work in the branch locally that you want to save:
```shell
git branch temp-save
```
2. Reset the branch to the position before the rebase (4dcbedcc76fe5fcb9cf1fab81a6431bc9f2e5b84):
```shell
git switch beginner-friendly-code-examples
git reset --hard 4dcbedc
```
3. Ensure that your local copy of main is up to date:
```shell
git switch main
git pull --rebase
git log -1
```
4. Ensure that your the log shows that your main branch is pointing at upstream/main (845b726e7d9bd871fa0667956d320a31b761fa38). If not, you probably have other code in your main branch (`git branch temp-main-save; git reset --hard upstream/main` to clear that)
5. Rebase the branch
```shell
git switch beginner-friendly-code-examples
git rebase main
```
6. If there are conflicts, resolve them with your mergetool of choice and once you're done `git rebase --continue`
7. Check that the log looks linear (i.e just the commits you're working on over the top of main)
```shell
git log
```
8. Push your changes to your remote fork
```shell
git push --force origin beginner-friendly-code-examples
```

**Comment by kdheepak** [comment]
I'm also happy to cherry-pick your commits and apply them in a separate PR and fix the conflicts, so you don't have to deal with this kind of git issues. And you'll still be credited for the commit if I cherry pick so git / github will show that it was your contribution. Let me know if you want me to resolve it that way.

**Comment by farmeroy** [comment]
SO, I wouldn't mind just giving it a shot to practice my git... how about I let you know how it goes after I give it another try?

**Comment by kdheepak** [comment]
Great! Thanks again for making the edits! We appreciate it!

**Comment by farmeroy** [comment]
Ok, I think I managed to rebase properly - the file diffs certainly look cleaner :smile:  Thanks so much for the lesson there, I think I know where I made my mistake (probably when I used `git add .` :face_with_spiral_eyes: )

File: src/tutorial/counter-app/event.md
Comment (kdheepak) [inline src/tutorial/counter-app/event.md]: If you modify `src/event.rs` for the `ratatui-counter-app` like so:

```
///// ANCHOR: event_import
use std::{
  sync::mpsc,
  thread,
  time::{Duration, Instant},
};

use anyhow::Result;
use crossterm::event::{self, Event as CrosstermEvent, KeyEvent, MouseEvent};
///// ANCHOR: event_import
```

then this block can become:

```suggestion
{{#include ./ratatui-counter-app/src/event.rs:event_import}}
```

File: src/tutorial/counter-app/multiple-files.md
Comment (kdheepak) [inline src/tutorial/counter-app/multiple-files.md]: I changed the tutorial such that you don't need `./src/lib.rs` anymore. And the recommendation is to add the `mod` directly to `main.rs`. 

I did this for a few reasons:

1. Most command line TUIs are not libraries and are just binary files, so there's no need for a `./src/lib.rs`.
2. Adding a `./src/lib.rs` makes testing slightly more annoying because now some tests will run twice (once when `cargo test` runs for `main.rs` and another when it runs it for `lib.rs`).
3. It's simpler to understand for someone new to rust and they don't have to worry about `lib` vs `bin` etc.

File: src/tutorial/counter-app/ratatui-counter-app/src/app.rs
Comment (kdheepak) [inline src/tutorial/counter-app/ratatui-counter-app/src/app.rs]: Use `super::*` instead.

```suggestion
mod tests {
  use super::*;
  #[test]
  fn test_app_increment_counter() {
    let mut app = App::default();
    app.increment_counter();
    assert_eq!(app.counter, 1);
  }

  #[test]
  fn test_app_decrement_counter() {
    let mut app = App::default();
    app.decrement_counter();
    assert_eq!(app.counter, 0);
  }
}
```

File: src/tutorial/counter-app/ratatui-counter-app/src/lib.rs
Comment (kdheepak) [inline src/tutorial/counter-app/ratatui-counter-app/src/lib.rs:1]: I recommend removing this file because it will be confusing. Refer to this comment for more details: https://github.com/ratatui-org/ratatui-book/pull/56/files#r1320677351

File: src/tutorial/counter-app/tui.md
Comment (kdheepak) [inline src/tutorial/counter-app/tui.md:31]: Add a `// --snip--` here before `}`.

```suggestion
  // --snip--
}
```

File: src/tutorial/counter-app/tui.md
Comment (kdheepak) [inline src/tutorial/counter-app/tui.md]: I think we want a `// --snip--` before and after the code block. Also, I think the `// --snip--` should be 2 spaces to align with the code included. And this should be `impl Tui`.

```suggestion
impl Tui {
  // --snip--
{{#include ./ratatui-counter-app/src/tui.rs:tui_exit}}
  // --snip--
}
```

File: src/tutorial/counter-app/ui.md
Comment (kdheepak) [inline src/tutorial/counter-app/ui.md]: Can we move this line after the code block below and remove the parenthesis?

File: src/tutorial/counter-app/multiple-files.md
Comment (farmeroy) [reply src/tutorial/counter-app/multiple-files.md]: Ah ok! That makes a lot of sense

Comment (farmeroy) [reply src/tutorial/counter-app/multiple-files.md]: I added a new ANCHOR for the mod imports in `main.rs` and reference that here now

File: src/tutorial/counter-app/ratatui-counter-app/src/app.rs
Comment (farmeroy) [reply src/tutorial/counter-app/ratatui-counter-app/src/app.rs:66]: This causes the counting-app to not compile (which is why I used `super` in line previously).
Taking the compiler's suggestion to do `use crate::app::App;` also results in a failure to compile. Maybe I'm doing something wrong?
```
 Compiling counting-app v0.1.0 (/home/raffaele/RustProjects/ratatui/counting-app)
warning: `counting-app` (lib) generated 1 warning (run `cargo fix --lib -p counting-app` to apply 1 suggestion)
error[E0433]: failed to resolve: use of undeclared type `App`
  --> src/app.rs:40:23
   |
40 |         let mut app = App::default();
   |                       ^^^ use of undeclared type `App`
   |
help: consider importing this struct
   |
38 +     use crate::app::App;
   |

error[E0433]: failed to resolve: use of undeclared type `App`
  --> src/app.rs:46:23
   |
46 |         let mut app = App::default();
   |                       ^^^ use of undeclared type `App`
   |
help: consider importing this struct
   |
38 +     use crate::app::App;
   |

For more information about this error, try `rustc --explain E0433`.
warning: `counting-app` (lib test) generated 1 warning (1 duplicate)
error: could not compile `counting-app` (lib test) due to 2 previous errors; 1 warning emitted
```

Comment (farmeroy) [reply src/tutorial/counter-app/ratatui-counter-app/src/app.rs]: Done

**Comment by kdheepak** [comment]
I think you've marked the conversations as resolved but haven't pushed your changes to this branch or this PR. Can you double check that?

**Comment by farmeroy** [comment]
Oh sorry! I've pushed them

File: src/tutorial/counter-app/tui.md
Comment (kdheepak) [reply src/tutorial/counter-app/tui.md:31]: I think you missed a `// --snip--` here.

**Comment by kdheepak** [comment]
This looks good to me! Do you want to make the PR as ready? I'm happy to merge after that.

**Review by kdheepak — APPROVED** [review]

**Comment by kdheepak** [comment]
Thanks for making the PR! These changes definitely improve the docs!

**Comment by farmeroy** [comment]
I'm really happy to! It's my first time contributing to open source software (if you couldn't tell lol) so it was a lot fun, thanks for being welcoming

**Comment by kdheepak** [comment]
We'd love for you to make more contributions if you are interested (even if just offer feedback) as you work through more of the tutorials and documentation.

**Pull request merged**
